pkg/sql/sem/tree/type_check.go

```diff
@@ -2238,6 +2239,27 @@ func typeCheckComparisonOp(
 			return nil, nil, nil, false, err
 		}
 		return typedLeft, typedRight, fn, false, nil
+
+	case leftIsTuple || rightIsTuple:
+		// Tuple must compare with a tuple type, as handled above.
+		typedLeft, err := foldedLeft.TypeCheck(ctx, semaCtx, types.Any)
+		if err != nil {
+			sigWithErr := fmt.Sprintf(compExprsFmt, left, op, right, err)
+			return nil, nil, nil, false,
+				pgerror.Newf(pgcode.InvalidParameterValue, unsupportedCompErrFmt, sigWithErr)
+		}
+		typedRight, err := foldedRight.TypeCheck(ctx, semaCtx, types.Any)
+		if err != nil {
+			sigWithErr := fmt.Sprintf(compExprsFmt, left, op, right, err)
+			return nil, nil, nil, false,
+				pgerror.Newf(pgcode.InvalidParameterValue, unsupportedCompErrFmt, sigWithErr)
+		}
+		if (typedLeft.ResolvedType().Family() == types.TupleFamily &&
+			typedRight.ResolvedType().Family() != types.TupleFamily) ||
+			(typedRight.ResolvedType().Family() == types.TupleFamily &&
+				typedLeft.ResolvedType().Family() != types.TupleFamily) {
+			handleTupleTypeMismatch = true
```

Comment:
If we made it into this case `leftIsTuple || rightIsTuple` don't we already know that `handleTupleTypeMismatch` should be true?

Reply:
> If we made it into this case `leftIsTuple || rightIsTuple` don't we already know that `handleTupleTypeMismatch` should be true?

That's what I originally thought, but `leftTuple` or `rightTuple` are only true when `left` or `right` are explicit `Tuple` structs:
https://github.com/cockroachdb/cockroach/blob/e67eb0ce76c76ff3d7afaa92f33f97f38da14a78/pkg/sql/sem/tree/expr.go#L818-L829

In some cases there is a scoped column with tuple type family on one side (which isn't a `Tuple` struct). Is it possible that both left and right could be scoped columns? Maybe. But the problem case here is a totally empty tuple, `()`, or `Row()` (cases with non-empty tuples error out through a different flow).  So the case of 2 scoped columns probably doesn't need to be handled, or if it does, we should wait until we have a case we can test before coding for it.

Reply:
Thanks for explaining!